# Project: No-show appoinments
## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
<li><a href="#eda">Exploratory Data Analysis</a></li>
<li><a href="#conclusions">Conclusions</a></li>
</ul>

<a id='intro'></a>
## 1 .Introduction


We will be working on the dataset 'No show appoinments' , This dataset collects information from 100k medical appointments in Brazil and is focused on the question of whether or not patients show up for their appointment.

In [1]:
#importing libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
pd.pandas.set_option('display.max_columns',None) 

C:\Users\sibin\Anaconda3\lib\site-packages\statsmodels\tools\_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


<a id='wrangling'></a>
## 2. Data Wrangling

#### 2.1 General Properties

In [2]:
#load the dataset
data = pd.read_csv("noshowappointments.csv")

In [3]:
# view dataframe
data.head(5)

PatientId  AppointmentID Gender          ScheduledDay  \
0  2.987250e+13        5642903      F  2016-04-29T18:38:08Z   
1  5.589978e+14        5642503      M  2016-04-29T16:08:27Z   
2  4.262962e+12        5642549      F  2016-04-29T16:19:04Z   
3  8.679512e+11        5642828      F  2016-04-29T17:29:31Z   
4  8.841186e+12        5642494      F  2016-04-29T16:07:23Z   

         AppointmentDay  Age      Neighbourhood  Scholarship  Hipertension  \
0  2016-04-29T00:00:00Z   62    JARDIM DA PENHA            0             1   
1  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             0   
2  2016-04-29T00:00:00Z   62      MATA DA PRAIA            0             0   
3  2016-04-29T00:00:00Z    8  PONTAL DE CAMBURI            0             0   
4  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             1   

   Diabetes  Alcoholism  Handcap  SMS_received No-show  
0         0           0        0             0      No  
1         0           0        0             0      No  
2         0           0        0             0      No  
3         0           0        0             0      No  
4         1           0        0             0      No

In [4]:
#inspect data
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110527 entries, 0 to 110526
Data columns (total 14 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   PatientId       110527 non-null  float64
 1   AppointmentID   110527 non-null  int64  
 2   Gender          110527 non-null  object 
 3   ScheduledDay    110527 non-null  object 
 4   AppointmentDay  110527 non-null  object 
 5   Age             110527 non-null  int64  
 6   Neighbourhood   110527 non-null  object 
 7   Scholarship     110527 non-null  int64  
 8   Hipertension    110527 non-null  int64  
 9   Diabetes        110527 non-null  int64  
 10  Alcoholism      110527 non-null  int64  
 11  Handcap         110527 non-null  int64  
 12  SMS_received    110527 non-null  int64  
 13  No-show         110527 non-null  object 
dtypes: float64(1), int64(8), object(5)
memory usage: 11.8+ MB


Data have 110527 rows and 14 columns, there are three datatypes: float, int, string .  
    'No-show' is taken as the dependent variable. We need to convert it to numerical to further ease up our analysis

In [5]:
data.describe()

PatientId  AppointmentID            Age    Scholarship  \
count  1.105270e+05   1.105270e+05  110527.000000  110527.000000   
mean   1.474963e+14   5.675305e+06      37.088874       0.098266   
std    2.560949e+14   7.129575e+04      23.110205       0.297675   
min    3.921784e+04   5.030230e+06      -1.000000       0.000000   
25%    4.172614e+12   5.640286e+06      18.000000       0.000000   
50%    3.173184e+13   5.680573e+06      37.000000       0.000000   
75%    9.439172e+13   5.725524e+06      55.000000       0.000000   
max    9.999816e+14   5.790484e+06     115.000000       1.000000   

        Hipertension       Diabetes     Alcoholism        Handcap  \
count  110527.000000  110527.000000  110527.000000  110527.000000   
mean        0.197246       0.071865       0.030400       0.022248   
std         0.397921       0.258265       0.171686       0.161543   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       4.000000   

        SMS_received  
count  110527.000000  
mean        0.321026  
std         0.466873  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         1.000000  
max         1.000000

Statistical observation on the dataset.

    Mean age of the patient is 37. There is an error in age value as minimum is shown ''-1', it needs to be corrected.
    20% of patients have Hypertension
    9 % of patients have a scholarship, that is they are enrolled in a welfare program
    32% people recieved an message reminding the appoinment
    
    

In [6]:
#check for duplicates.
data.duplicated().value_counts()

False    110527
dtype: int64

Data has zero duplicate rows.

##### Handling missing values

In [7]:
data.isnull().sum()

PatientId         0
AppointmentID     0
Gender            0
ScheduledDay      0
AppointmentDay    0
Age               0
Neighbourhood     0
Scholarship       0
Hipertension      0
Diabetes          0
Alcoholism        0
Handcap           0
SMS_received      0
No-show           0
dtype: int64

the data is said to have zero missing values

#### 2.2 Cleaning data

In [8]:
x= data[data['Age'] == -1]
x

PatientId  AppointmentID Gender          ScheduledDay  \
99832  4.659432e+14        5775010      F  2016-06-06T08:58:13Z   

             AppointmentDay  Age Neighbourhood  Scholarship  Hipertension  \
99832  2016-06-06T00:00:00Z   -1         ROMÃO            0             0   

       Diabetes  Alcoholism  Handcap  SMS_received No-show  
99832         0           0        0             0      No

There is one column with age = -1, this is an outlier. So we will delete the row with age = -1

In [9]:
# removing the row with age = -1
data = data[data.Age != -1]

In [10]:
#patientID and AppoinmentID are not found usedfull, so drop them
data.drop(['PatientId','AppointmentID'], axis = 1, inplace = True)

In [11]:
# schedule day and appointment day have time associated with data, let us remove that and make the data more meaningfull
#split column by T 
colu = ['ScheduledDay' , 'AppointmentDay']
for col in colu:
    data[col] = data[col].apply(lambda x : x.split('T')[0])

In [12]:
data['ScheduledDay'] = pd.to_datetime(data['ScheduledDay'])
data['AppointmentDay'] = pd.to_datetime(data['AppointmentDay'])

In [13]:
data.head()

Gender ScheduledDay AppointmentDay  Age      Neighbourhood  Scholarship  \
0      F   2016-04-29     2016-04-29   62    JARDIM DA PENHA            0   
1      M   2016-04-29     2016-04-29   56    JARDIM DA PENHA            0   
2      F   2016-04-29     2016-04-29   62      MATA DA PRAIA            0   
3      F   2016-04-29     2016-04-29    8  PONTAL DE CAMBURI            0   
4      F   2016-04-29     2016-04-29   56    JARDIM DA PENHA            0   

   Hipertension  Diabetes  Alcoholism  Handcap  SMS_received No-show  
0             1         0           0        0             0      No  
1             0         0           0        0             0      No  
2             0         0           0        0             0      No  
3             0         0           0        0             0      No  
4             1         1           0        0             0      No

In [14]:
data["Appoinment_Gap"] = data['AppointmentDay'].dt.date - data['ScheduledDay'].dt.date
data['Appoinment_Gap'] = data['Appoinment_Gap'].apply(lambda x : x.days)

In [15]:
data.head()
data['Appoinment_Gap'].describe()

count    110526.000000
mean         10.183794
std          15.255034
min          -6.000000
25%           0.000000
50%           4.000000
75%          15.000000
max         179.000000
Name: Appoinment_Gap, dtype: float64

In [16]:
# minimum value of appoinment gap us -6, which cannot be true, needs further analyis
x = data[data['Appoinment_Gap']< 0]
x

Gender ScheduledDay AppointmentDay  Age  Neighbourhood  Scholarship  \
27033      M   2016-05-10     2016-05-09   38    RESISTÊNCIA            0   
55226      F   2016-05-18     2016-05-17   19  SANTO ANTÔNIO            0   
64175      F   2016-05-05     2016-05-04   22     CONSOLAÇÃO            0   
71533      F   2016-05-11     2016-05-05   81  SANTO ANTÔNIO            0   
72362      M   2016-05-04     2016-05-03    7     TABUAZEIRO            0   

       Hipertension  Diabetes  Alcoholism  Handcap  SMS_received No-show  \
27033             0         0           0        1             0     Yes   
55226             0         0           0        1             0     Yes   
64175             0         0           0        0             0     Yes   
71533             0         0           0        0             0     Yes   
72362             0         0           0        0             0     Yes   

       Appoinment_Gap  
27033              -1  
55226              -1  
64175              -1  
71533              -6  
72362              -1

SceduledDay is the day someone called or registered the appointment, , and AppointmentDay is The day of the actuall appointment, when they have to visit the doctor.       
 AppoinmentDay cannot be before scheduled day, so the rows with negative appoinment gap will be considered as an outlier  

In [17]:
# removing the row with appoinment gap = -1 and -6
data = data[data.Appoinment_Gap != -1 ]

In [ ]:

data['Appoinment_Gap'].describe()

No-show is a categorical variable, it is best if it can be transoformed to numerical variable as it helps in prediction.
No-show 'Yes' :1
        'No' : 0
        

In [19]:
data['No-show'] = data['No-show'].apply(lambda x : 1 if x == 'Yes'else 0)

In [20]:
data.head()

Gender ScheduledDay AppointmentDay  Age      Neighbourhood  Scholarship  \
0      F   2016-04-29     2016-04-29   62    JARDIM DA PENHA            0   
1      M   2016-04-29     2016-04-29   56    JARDIM DA PENHA            0   
2      F   2016-04-29     2016-04-29   62      MATA DA PRAIA            0   
3      F   2016-04-29     2016-04-29    8  PONTAL DE CAMBURI            0   
4      F   2016-04-29     2016-04-29   56    JARDIM DA PENHA            0   

   Hipertension  Diabetes  Alcoholism  Handcap  SMS_received  No-show  \
0             1         0           0        0             0        0   
1             0         0           0        0             0        0   
2             0         0           0        0             0        0   
3             0         0           0        0             0        0   
4             1         1           0        0             0        0   

   Appoinment_Gap  
0               0  
1               0  
2               0  
3               0  
4               0

<a id='wrangling'></a>
## 2. Data Wrang